client.go

```diff
@@ -1,137 +1,185 @@
-	filter := SearchFilter{
-		Operator: "AND",
-		Terms: []FilterTerm{{
-			Name:  "metadata._gcvi_tags",
-			Value: tag,
-		}},
-	}
-	schema := SearchCriteriaSchema{
-		DocTypes: []string{"assets"},
-		Filter:   filter,
+	req, err := c.authRequest(http.MethodGet, apiPath, body, header)
+	if err != nil {
+		return nil, err
 	}
-	body, err := json.Marshal(schema)
-	return body, err
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
 }
 
-func makeSearchRequest(apiUrl string, header http.Header, body []byte) (*http.Request, error) {
-	url, err := url.Parse(apiUrl)
+func (c *IClient) authPost(apiPath string, body io.Reader, header http.Header) (*http.Response, error) {
+
+	req, err := c.authRequest(http.MethodPost, apiPath, body, header)
 	if err != nil {
 		return nil, err
 	}
-	req := &http.Request{
-		Method: http.MethodPost,
-		URL:    url,
-		Header: header,
-		Body:   ioutil.NopCloser(bytes.NewReader(body)),
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
+}
+
+// Looks for an error message in the response body and parses it into a
+// IError object
+func (c *IClient) parseError(body []byte) error {
+	err := &IError{}
+	if json.Unmarshal(body, err) != nil {
+		return nil
 	}
-	return req, nil
+	return err
 }
 
-func main() {
-	flag.StringVar(&tag, "tag", "GPTeaching", "Tag to search for all GPTeaching assets")
-	client := IconikClient{client: &http.Client{}}
-	header := makeSearchHeader()
-	body, err := makeSearchBody(tag)
+// Attempts to parse a response body into the provided result struct
+func (c *IClient) parseResponse(resp *http.Response, result interface{}) error {
+	defer resp.Body.Close()
+
+	// Read response body
+	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		fmt.Errorf("makeSearchBody() failed: %v", err)
-		return
+		return err
 	}
-	req, err := makeSearchRequest(searchUrl, header, body)
+
+	if c.Debug {
+		log.Printf("Response: %s", body)
+	}
+
+	// Check response code
+	switch resp.StatusCode {
+	case 200: // Response is OK
+	case 401:
+		if err := c.parseError(body); err != nil {
+			return err
+		}
+		return &IError{
+			Errors: []string{"UNKNOWN; error message not parsable"},
+		}
+	default:
+		if err := c.parseError(body); err != nil {
+			return err
+		}
+		return &IError{
+			Errors: []string{"UNKNOWN; error message not parsable"},
+		}
+	}
+
+	return json.Unmarshal(body, result)
+}
+
+func (c *IClient) makeSearchHeader() http.Header {
+	header := make(http.Header)
+	header.Add("accept", "application/json")
+	header.Add("App-Id", c.AppID)
+	header.Add("Content-Type", "application/json")
+	header.Add("Auth-Token", c.Token)
+	return header
+}
+
+// ApiRequest performs an Iconik API request.
+// Args:
+// apiPath: The API Resource
+// Request The filled request object (eg. SearchCriteriaSchema) to be used.
+// Response: The response object (eg. SearchResponse) to be filled out.
+func (c *IClient) ApiRequest(apiPath string, request interface{}, response interface{}) error {
```

Comment:
So, we want to export the specific calls we are making. So I would expect us to export "SearchWithTags()" and "GetPresignedDownloadUrl(assetID)" and we should NOT export a generic ApiRequest(... interface{}, interface{})

some points:
- people using this client should not need to know apiPaths, that is why they want to use a client
- we shouldn't take in nor return interface{} types unless truly the API returns things of random types. In our case, the API takes specific types and returns specific types, so we should be as specific as we can. Else users will be having to figure out what to pass in or what is returned, and that is what the client is for, so users don't have to figure that out.
- why are you taking in a response object rather than return it?

Reply:
Exported SearchWithTags! GetPresignedDownloadID will be in the follow-up CL.

@ taking response object: done!